### PR TITLE
Subaru: gen2 long safety

### PR DESCRIPTION
**Pull request by jnewb1**
Added more restrictive testing around the UDS requests (only tester present and RDBI for the exact button data we need are allowed)

### CODE REVIEW

**Comment by sshane** [comment]
original PR: https://github.com/commaai/panda/pull/1587

File: tests/safety/test_subaru.py [hunk 1]
```diff
@@ -195,7 +206,43 @@ class TestSubaruGen2TorqueStockLongitudinalSafety(TestSubaruStockLongitudinalSaf
...
+  MAX_RATE_DOWN = 40
+  MAX_TORQUE = 1000
+
+  FLAGS = Panda.FLAG_SUBARU_LONG | Panda.FLAG_SUBARU_GEN2
+  TX_MSGS = lkas_tx_msgs(SUBARU_ALT_BUS) + long_tx_msgs(SUBARU_ALT_BUS) + additional_tx_msgs(SUBARU_ALT_BUS)
+
+  def _rdbi_msg(self, did: int):
+    return b'\x03\x22' + did.to_bytes(2) + b'\x00\x00\x00\x00'
+
+  def _es_uds_msg(self, msg):
```
Comment (sshane) [inline tests/safety/test_subaru.py:226]:
```suggestion
  def _es_uds_msg(self, msg: bytes):
```

File: tests/safety/test_subaru.py [hunk 2]
```diff
@@ -195,7 +206,43 @@ class TestSubaruGen2TorqueStockLongitudinalSafety(TestSubaruStockLongitudinalSaf
...
+    return libpanda_py.make_CANPacket(MSG_SUBARU_ES_UDS_Request, 2, msg)
+
+  def test_es_uds_message(self):
+    tester_present = b'\x02\x3E\x80\x00\x00\x00\x00\x00'
+
+    button_did = 0x1130
+
+    # Tester present is allowed for gen2 long to keep eyesight disabled
+    self.assertTrue(self._tx(self._es_uds_msg(tester_present)))
```
Comment (sshane) [inline tests/safety/test_subaru.py:235]: add an assertFalse (check honda for an example)

File: tests/safety/test_subaru.py (hunk 1 above)
Comment (jnewb1) [reply tests/safety/test_subaru.py]: done

File: tests/safety/test_subaru.py (hunk 2 above)
Comment (jnewb1) [reply tests/safety/test_subaru.py:212]: done, although I did go through all the SID's at the bottom

File: board/safety/safety_subaru.h [hunk 3]
```diff
@@ -39,20 +39,32 @@ const LongitudinalLimits SUBARU_LONG_LIMITS = {
...
+  {MSG_SUBARU_ES_Infotainment,   SUBARU_MAIN_BUS, 8}, \
+
+#define SUBARU_COMMON_LONG_TX_MSGS(alt_bus)           \
+  {MSG_SUBARU_ES_Brake,          alt_bus,         8}, \
+  {MSG_SUBARU_ES_Status,         alt_bus,         8}, \
 
-#define SUBARU_COMMON_LONG_TX_MSGS(alt_bus)         \
-  {MSG_SUBARU_ES_Brake,        SUBARU_MAIN_BUS, 8}, \
-  {MSG_SUBARU_ES_Status,       SUBARU_MAIN_BUS, 8}, \
+#define SUBARU_GEN2_LONG_ADDITIONAL_TX_MSGS(alt_bus)  \
```
Comment (sshane) [inline board/safety/safety_subaru.h:63]: any reason to specify alt bus since it's only for gen 2?

File: board/safety/safety_subaru.h [hunk 4]
```diff
@@ -219,6 +240,16 @@ static int subaru_tx_hook(CANPacket_t *to_send) {
     violation |= longitudinal_transmission_rpm_checks(transmission_rpm, SUBARU_LONG_LIMITS);
   }
 
+  if (addr == MSG_SUBARU_ES_UDS_Request){
```
Comment (sshane) [inline board/safety/safety_subaru.h:243]:
```suggestion
  if (addr == MSG_SUBARU_ES_UDS_Request) {
```

File: tests/safety/test_subaru.py [hunk 5]
```diff
@@ -7,35 +7,46 @@
...
+          [MSG_SUBARU_ES_Distance,       alt_bus],
+          [MSG_SUBARU_ES_DashStatus,     SUBARU_MAIN_BUS],
+          [MSG_SUBARU_ES_LKAS_State,     SUBARU_MAIN_BUS],
+          [MSG_SUBARU_ES_Infotainment,   SUBARU_MAIN_BUS]]
+
+def long_tx_msgs(alt_bus):
+  return [[MSG_SUBARU_ES_Brake,          alt_bus],
+          [MSG_SUBARU_ES_Status,         alt_bus]]
+
+def additional_tx_msgs(alt_bus):
```
Comment (sshane) [inline tests/safety/test_subaru.py:45]: can you call it the same thing as in panda safety (gen2) if it is? same comment about `alt_bus` argument too

**Review by sshane — APPROVED** [review]

File: board/safety/safety_subaru.h (hunk 3 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: removed

File: board/safety/safety_subaru.h (hunk 4 above)
Comment (jnewb1) [reply board/safety/safety_subaru.h]: done

File: tests/safety/test_subaru.py (hunk 5 above)
Comment (jnewb1) [reply tests/safety/test_subaru.py]: done

**Comment by jnewb1** [comment]
@adeebshihadeh want to do a final review here?

File: tests/safety/test_subaru.py [hunk 6]
```diff
@@ -171,7 +182,47 @@ class TestSubaruGen2TorqueStockLongitudinalSafety(TestSubaruStockLongitudinalSaf
...
+  TX_MSGS = lkas_tx_msgs(SUBARU_MAIN_BUS) + long_tx_msgs(SUBARU_MAIN_BUS)
+
+
+class TestSubaruGen2LongitudinalSafety(TestSubaruLongitudinalSafetyBase, TestSubaruTorqueSafetyBase):
+  ALT_MAIN_BUS = SUBARU_ALT_BUS
+  ALT_CAM_BUS = SUBARU_ALT_BUS
+
+  MAX_RATE_UP = 40
+  MAX_RATE_DOWN = 40
+  MAX_TORQUE = 1000
```
Comment (adeebshihadeh) [inline tests/safety/test_subaru.py:194]: why are these copied?

File: tests/safety/test_subaru.py [hunk 7]
```diff
@@ -7,35 +7,46 @@
...
-MSG_SUBARU_ES_LKAS          = 0x122
-MSG_SUBARU_ES_LKAS_ANGLE    = 0x124
-MSG_SUBARU_ES_Brake         = 0x220
-MSG_SUBARU_ES_Distance      = 0x221
-MSG_SUBARU_ES_Status        = 0x222
-MSG_SUBARU_ES_DashStatus    = 0x321
-MSG_SUBARU_ES_LKAS_State    = 0x322
-MSG_SUBARU_ES_Infotainment  = 0x323
+MSG_SUBARU_Brake_Status      = 0x13c
+MSG_SUBARU_CruiseControl     = 0x240
```
Comment (adeebshihadeh) [inline tests/safety/test_subaru.py:11]: these names are super long, you can properly namespace here, so they don't need to be that long

File: tests/safety/test_subaru.py (hunk 6 above)
Comment (jnewb1) [reply tests/safety/test_subaru.py]: made a base class

File: tests/safety/test_subaru.py (hunk 7 above)
Comment (jnewb1) [reply tests/safety/test_subaru.py]: scoped in an enum

**Pull request merged**